### PR TITLE
Refactor package `internal/attribute` to not use generics.

### PR DESCRIPTION
**Pull request by Petrie**
Signed-off-by: Peter Liu <lpfvip2008@gmail.com>
resolve https://github.com/open-telemetry/opentelemetry-go/issues/3723

### CODE REVIEW

File: internal/attribute/attribute.go [hunk 1]
```diff
@@ -22,24 +22,90 @@ import (
 	"reflect"
 )
 
-// SliceValue convert a slice into an array with same elements as slice.
-func SliceValue[T bool | int64 | float64 | string](v []T) any {
-	var zero T
+// SliceValue convert a bool slice into an array with same elements as slice.
```
Comment (dmathieu) [inline internal/attribute/attribute.go:25]:
```suggestion
// BoolSliceValue converts a bool slice into an array with same elements as slice.
```

File: internal/attribute/attribute.go [hunk 2]
```diff
@@ -22,24 +22,90 @@ import (
...
+	var zero bool
 	cp := reflect.New(reflect.ArrayOf(len(v), reflect.TypeOf(zero)))
-	copy(cp.Elem().Slice(0, len(v)).Interface().([]T), v)
+	copy(cp.Elem().Slice(0, len(v)).Interface().([]bool), v)
 	return cp.Elem().Interface()
 }
 
-// AsSlice convert an array into a slice into with same elements as array.
-func AsSlice[T bool | int64 | float64 | string](v any) []T {
+// SliceValue convert a int64 slice into an array with same elements as slice.
```
Comment (dmathieu) [inline internal/attribute/attribute.go:33]:
```suggestion
// Int64SliceValue converts a int64 slice into an array with same elements as slice.
```

File: internal/attribute/attribute.go [hunk 3]
```diff
@@ -22,24 +22,90 @@ import (
...
+func Int64SliceValue(v []int64) any {
+	var zero int64
+	cp := reflect.New(reflect.ArrayOf(len(v), reflect.TypeOf(zero)))
+	copy(cp.Elem().Slice(0, len(v)).Interface().([]int64), v)
+	return cp.Elem().Interface()
+}
+
+// SliceValue convert a float64 slice into an array with same elements as slice.
```
Comment (dmathieu) [inline internal/attribute/attribute.go:41]:
```suggestion
// Float64SliceValue converts a float64 slice into an array with same elements as slice.
```

File: internal/attribute/attribute.go [hunk 4]
```diff
@@ -22,24 +22,90 @@ import (
...
+func Float64SliceValue(v []float64) any {
+	var zero float64
+	cp := reflect.New(reflect.ArrayOf(len(v), reflect.TypeOf(zero)))
+	copy(cp.Elem().Slice(0, len(v)).Interface().([]float64), v)
+	return cp.Elem().Interface()
+}
+
+// SliceValue convert a string slice into an array with same elements as slice.
```
Comment (dmathieu) [inline internal/attribute/attribute.go:49]:
```suggestion
// StringSliceValue converts a string slice into an array with same elements as slice.
```

File: internal/attribute/attribute.go [hunk 5]
```diff
@@ -22,24 +22,90 @@ import (
...
+func StringSliceValue(v []string) any {
+	var zero string
+	cp := reflect.New(reflect.ArrayOf(len(v), reflect.TypeOf(zero)))
+	copy(cp.Elem().Slice(0, len(v)).Interface().([]string), v)
+	return cp.Elem().Interface()
+}
+
+// AsSlice convert an bool array into a slice into with same elements as array.
```
Comment (dmathieu) [inline internal/attribute/attribute.go:57]:
```suggestion
// AsBoolSlice converts a bool array into a slice into with same elements as array.
```

File: internal/attribute/attribute.go [hunk 6]
```diff
@@ -22,24 +22,90 @@ import (
...
+	}
+	var zero bool
+	correctLen := rv.Len()
+	correctType := reflect.ArrayOf(correctLen, reflect.TypeOf(zero))
+	cpy := reflect.New(correctType)
+	_ = reflect.Copy(cpy.Elem(), rv)
+	return cpy.Elem().Slice(0, correctLen).Interface().([]bool)
+}
+
+// AsSlice convert an int64 array into a slice into with same elements as array.
```
Comment (dmathieu) [inline internal/attribute/attribute.go:71]:
```suggestion
// AsInt64Slice converts an int64 array into a slice into with same elements as array.
```

File: internal/attribute/attribute.go [hunk 7]
```diff
@@ -22,24 +22,90 @@ import (
...
+	}
+	var zero int64
+	correctLen := rv.Len()
+	correctType := reflect.ArrayOf(correctLen, reflect.TypeOf(zero))
+	cpy := reflect.New(correctType)
+	_ = reflect.Copy(cpy.Elem(), rv)
+	return cpy.Elem().Slice(0, correctLen).Interface().([]int64)
+}
+
+// AsSlice convert an float64 array into a slice into with same elements as array.
```
Comment (dmathieu) [inline internal/attribute/attribute.go:85]:
```suggestion
// AsFloat64Slice converts a float64 array into a slice into with same elements as array.
```

File: internal/attribute/attribute.go [hunk 8]
```diff
@@ -22,24 +22,90 @@ import (
...
+	}
+	var zero float64
+	correctLen := rv.Len()
+	correctType := reflect.ArrayOf(correctLen, reflect.TypeOf(zero))
+	cpy := reflect.New(correctType)
+	_ = reflect.Copy(cpy.Elem(), rv)
+	return cpy.Elem().Slice(0, correctLen).Interface().([]float64)
+}
+
+// AsSlice convert an string array into a slice into with same elements as array.
```
Comment (dmathieu) [inline internal/attribute/attribute.go:99]:
```suggestion
// AsStringSlice converts a string array into a slice into with same elements as array.
```

File: internal/attribute/attribute_test.go [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+// limitations under the License.
+
+package attribute
+
+import (
+	"reflect"
+	"testing"
+)
+
+func TestBoolSliceValue(t *testing.T) {
```
Comment (dmathieu) [inline internal/attribute/attribute_test.go:22]: All these tests are doing exactly the same thing, and they use subtests but all have a single one.
That's seems either a lot of repetition, or unneeded complexity.

How about having a single test, which tests all slice values, and allows a func parameter which calls the proper method and returns an interface with the value to be compared?

**Review by dmathieu — COMMENTED** [review]
Thank you! ❤️ 
I believe this will also need a changelog entry.

File: internal/attribute/attribute_test.go (hunk 9 above)
Comment (Petrie) [reply internal/attribute/attribute_test.go]: @dmathieu 
Thank you so much for your suggestion.

Unit tests have been refactored, please take a look.

File: CHANGELOG.md [hunk 10]
```diff
@@ -31,6 +31,10 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
     - `OtelLibraryVersion` -> `OTelLibraryVersion`
     - `OtelStatusDescription` -> `OTelStatusDescription`
 
+### Changed
```
Comment (dmathieu) [inline CHANGELOG.md:34]: I'd have put that as `Fixed`, as it fixes a known bug.

**Review by dmathieu — APPROVED** [review]

File: internal/attribute/attribute.go [hunk 11]
```diff
@@ -22,24 +22,90 @@ import (
...
+	var zero bool
 	cp := reflect.New(reflect.ArrayOf(len(v), reflect.TypeOf(zero)))
-	copy(cp.Elem().Slice(0, len(v)).Interface().([]T), v)
+	copy(cp.Elem().Slice(0, len(v)).Interface().([]bool), v)
 	return cp.Elem().Interface()
 }
 
-// AsSlice convert an array into a slice into with same elements as array.
-func AsSlice[T bool | int64 | float64 | string](v any) []T {
+// Int64SliceValue converts a int64 slice into an array with same elements as slice.
```
Comment (MrAlias) [inline internal/attribute/attribute.go:33]:
```suggestion
// Int64SliceValue converts an int64 slice into an array with same elements as slice.
```

File: internal/attribute/attribute_test.go [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+// limitations under the License.
+
+package attribute
+
+import (
+	"reflect"
+	"testing"
+)
+
+var wrapFloat64SliceValue = func(v any) any {
```
Comment (MrAlias) [inline internal/attribute/attribute_test.go:22]: `any` was introduced in Go 1.18, `interface{}` should be used instead.

Comment (Petrie) [reply internal/attribute/attribute_test.go]: fixed

**Review by MrAlias — APPROVED** [review]

**Review by hanyuancheung — APPROVED** [review]

**Pull request merged**